api/manifest.go

```diff
@@ -40,58 +38,98 @@ func apiManifestHandler(w http.ResponseWriter, r *http.Request) {
 	w.Write(manifest)
 }
 
-func getManifest(ctx context.Context, manifestAPI manifest.API, sha string, paths []string) (string, []byte, error) {
-	if sha == "" {
-		sha = "latest"
-	}
-	fetchedSHA := sha
+func getManifest(log shared.Logger, manifestAPI manifest.API, sha string, paths []string) (string, []byte, error) {
+	mc := manifestAPI.NewMemcache(time.Hour * 48)
+	// Shorter expiry for latest SHA, to keep it current.
+	latestMC := manifestAPI.NewMemcache(time.Minute * 5)
+
 	var body []byte
-	cached, err := memcache.Get(ctx, manifestCacheKey(sha))
 
-	if err != nil && err != memcache.ErrCacheMiss {
-		return "", nil, err
-	} else if cached != nil {
-		// "latest" caches which SHA is latest; Return the manifest for that SHA.
-		if sha == "latest" {
-			return getManifest(ctx, manifestAPI, string(cached.Value), paths)
+	if shared.IsLatest(sha) {
+		// Attempt to find the "latest" SHA in cache.
+		latestSHA, err := readByKey(latestMC, "latest")
+		if err == nil {
+			// Found! Now delegate to get manifest for that specific SHA.
+			return getManifest(log, manifestAPI, string(latestSHA), paths)
 		}
-		body = cached.Value
+		log.Debugf("Latest SHA not found in cache: %v", err)
 	} else {
-		fetchedSHA, body, err = manifestAPI.GetManifestForSHA(sha)
+		// Attempt to find the manifest for a specific SHA in cache.
+		var err error
+		body, err = readByKey(mc, sha)
 		if err != nil {
-			return fetchedSHA, nil, err
-		}
-		if paths != nil {
-			if body, err = manifest.Filter(body, paths); err != nil {
-				return fetchedSHA, nil, err
-			}
+			log.Debugf("Manifest for SHA %s not found in cache: %v", sha, err)
 		}
+		// Do not return here yet as we still need to filter by paths.
 	}
 
-	item := &memcache.Item{
-		Key:   manifestCacheKey(fetchedSHA),
-		Value: body,
+	var fetchedSHA string
+	if body != nil {
+		// Cache hit; fetchedSHA is requested SHA, which is guaranteed to be specific.
+		fetchedSHA = sha
+	} else {
+		// Cache missed; download the manifest for real.
+		var err error
+		if fetchedSHA, body, err = manifestAPI.GetManifestForSHA(sha); err != nil {
+			return fetchedSHA, nil, err
+		}
+
+		// Write manifest to cache.
+		if err := writeByKey(mc, fetchedSHA, body); err != nil {
+			log.Errorf("Error writing manifest to cache: %v", err)
+		}
 	}
-	memcache.Set(ctx, item)
 
-	// Shorter expiry for latest SHA, to keep it current.
+	// Write latest SHA to cache, if needed.
 	if shared.IsLatest(sha) {
-		latestSHAItem := &memcache.Item{
-			Key:        manifestCacheKey("latest"),
-			Value:      []byte(fetchedSHA),
-			Expiration: time.Minute * 5,
+		if err := writeByKey(latestMC, "latest", []byte(fetchedSHA)); err != nil {
+			log.Errorf("Error writing latest SHA to cache: %v", err)
 		}
-		memcache.Set(ctx, latestSHAItem)
 	}
 
+	// Decompress the manifest and filter it by paths if needed.
 	gzReader, err := gzip.NewReader(bytes.NewReader(body))
```

Comment:
I think this needs to be closed; see [doc](https://golang.org/pkg/compress/gzip/#NewReader) and [an example](https://github.com/web-platform-tests/wpt.fyi/blob/master/shared/metadata_util.go#L74)